### PR TITLE
Tx lookup skip

### PR DESCRIPTION
**Pull request by ceyonur**
It modifies type of `TxLookupLimit` flag from `uint64` type to `int64` with support for negative numbers which indicates to skip tx indexing.

Using negative number also means stop the tx unindexer from removing old indices. This is because unindexer still writes tail to DB and tries to keep up with the latest tip (even though the actual tx indexing is not performed on DB). Stopping (

Closes: #397

### CODE REVIEW

File: plugin/evm/config.go [hunk 1]
```diff
@@ -193,9 +196,10 @@ type Config struct {
 
 	// TxLookupLimit is the maximum number of blocks from head whose tx indices
 	// are reserved:
+	// 	* -1:  means skip tx indexing
```
Comment (patrick-ogrady) [inline plugin/evm/config.go:199]: Should also clarify that no txs will be cleaned up in this mode here.

File: eth/ethconfig/config.go [hunk 2]
```diff
@@ -153,7 +153,8 @@ type Config struct {
 
 	// TxLookupLimit is the maximum number of blocks from head whose tx indices
 	// are reserved:
+	// 	* -1:  means skip tx indexing
```
Comment (aaronbuchwald) [inline eth/ethconfig/config.go:156]: nit: align formatting with the values below

**Review by aaronbuchwald — APPROVED** [review]

File: core/blockchain.go [hunk 3]
```diff
@@ -413,7 +413,10 @@ func NewBlockChain(
 // unindexBlocks unindexes transactions depending on user configuration
 func (bc *BlockChain) unindexBlocks(tail uint64, head uint64, done chan struct{}) {
 	start := time.Now()
-	txLookupLimit := bc.cacheConfig.TxLookupLimit
+	if bc.cacheConfig.TxLookupLimit < 0 {
```
Comment (patrick-ogrady) [inline core/blockchain.go:416]: Because `unindexBlocks` is spawned by `dispatchTxUnindexer`, do we need to worry about this?

File: core/blockchain.go [hunk 4]
```diff
@@ -431,7 +434,10 @@ func (bc *BlockChain) unindexBlocks(tail uint64, head uint64, done chan struct{}
 // Meaning that this function should never be called.
 func (bc *BlockChain) dispatchTxUnindexer() {
 	defer bc.wg.Done()
-	txLookupLimit := bc.cacheConfig.TxLookupLimit
+	if bc.cacheConfig.TxLookupLimit < 0 {
```
Comment (patrick-ogrady) [inline core/blockchain.go:437]: Because we check `TxLookupLimit` before spawning `dispatchTxUnindexer`, do we need to worry about this?

**Comment by patrick-ogrady** [comment]
This is very close. I think it is functionally correct right now (just has unnecessary checks).

File: core/blockchain.go (hunk 3 above)
Comment (ceyonur) [reply core/blockchain.go]: yup correct. I put it to be extra safe (in case this is directly called by tests), but I guess it's fine to just add it as a comment.

File: core/blockchain.go (hunk 4 above)
Comment (ceyonur) [reply core/blockchain.go]: responded: https://github.com/ava-labs/coreth/pull/400#discussion_r1400405969

File: core/blockchain_test.go [hunk 5]
```diff
@@ -706,14 +729,16 @@ func TestTransactionIndices(t *testing.T) {
...
 	lastAcceptedHash := chain.CurrentHeader().Hash()
 
 	// Reconstruct a block chain which only reserves limited tx indices
 	// 128 blocks were previously indexed. Now we add a new block at each test step.
-	limit := []uint64{130 /* 129 + 1 reserve all */, 64 /* drop stale */, 32 /* shorten history */}
-	tails := []uint64{0 /* reserve all */, 67 /* 130 - 64 + 1 */, 100 /* 131 - 32 + 1 */}
+	limit := []int64{130 /* 129 + 1 reserve all */, 64 /* drop stale */, 32 /* shorten history */, -1 /* skip indexing */, -1 /* skip indexing */}
+	tails := []uint64{0 /* reserve all */, 67 /* 130 - 64 + 1 */, 100 /* 131 - 32 + 1 */, 100 /* skip indexing */, 100 /* skip indexing */}
+	// lastIndexed should stop with limit=-1 otherwise it should be the current block number
+	lastIndexed := []uint64{129 /* current block number */, 130 /* current block number */, 131 /* current block number */, 131 /* skip indexing */, 131 /* skip indexing */}
```
Comment (aaronbuchwald) [inline core/blockchain_test.go:741]: we should convert this to a struct in the future. It was a bit hard to follow before and going from 2 to 3 values makes it a bit harder to follow.

File: core/blockchain.go [hunk 6]
```diff
@@ -160,7 +160,7 @@ type CacheConfig struct {
 	SnapshotVerify                  bool          // Verify generated snapshots
 	Preimages                       bool          // Whether to store preimage of trie key to the disk
 	AcceptedCacheSize               int           // Depth of accepted headers cache and accepted logs cache at the accepted tip
-	TxLookupLimit                   uint64        // Number of recent blocks for which to maintain transaction lookup indices
+	TxLookupLimit                   int64         // Number of recent blocks for which to maintain transaction lookup indices
```
Comment (darioush) [inline core/blockchain.go:163]: Should we add a new parameter like`tx-indexing-disabled` instead of adding -1?
Seems this would simplify validation

File: core/blockchain.go [hunk 7]
```diff
@@ -405,17 +405,18 @@ func NewBlockChain(
...
-	if bc.cacheConfig.TxLookupLimit != 0 {
+	if bc.cacheConfig.TxLookupLimit > 0 {
 		bc.wg.Add(1)
 		go bc.dispatchTxUnindexer()
 	}
 	return bc, nil
 }
 
 // unindexBlocks unindexes transactions depending on user configuration
+// Invariant: This should not be called if TxLookupLimit is negative.
```
Comment (darioush) [inline core/blockchain.go:416]: would it be clearer to take a uint param here & to dispatchTxUnindexer?

File: core/blockchain.go [hunk 8]
```diff
@@ -485,7 +487,10 @@ func (bc *BlockChain) dispatchTxUnindexer() {
 // - updating the acceptor tip index
 func (bc *BlockChain) writeBlockAcceptedIndices(b *types.Block) error {
 	batch := bc.db.NewBatch()
-	rawdb.WriteTxLookupEntriesByBlock(batch, b)
+	// -1 means skip writing tx lookup entries
+	if bc.cacheConfig.TxLookupLimit != -1 {
```
Comment (darioush) [inline core/blockchain.go:491]: Let's be consistent with using < 0 here

File: core/blockchain_test.go [hunk 9]
```diff
@@ -749,7 +775,7 @@ func TestTransactionIndices(t *testing.T) {
 // TestCanonicalHashMarker tests all the canonical hash markers are updated/deleted
 // correctly in case reorg is called.
 func TestCanonicalHashMarker(t *testing.T) {
-	var cases = []struct {
+	cases := []struct {
```
Comment (darioush) [inline core/blockchain_test.go:778]: could we revert the unnecessary changes?

File: core/blockchain_test.go (hunk 5 above)
Comment (darioush) [reply core/blockchain_test.go:750]: might be nice to do this as part of this PR (or immediate follow-up), doesn't seem like a big change.

File: core/blockchain.go (hunk 6 above)
Comment (ceyonur) [reply core/blockchain.go:163]: In that case we should validate it against TxLookupLimit and it might be confusing to use both of them at the same time.

File: core/blockchain.go (hunk 8 above)
Comment (ceyonur) [reply core/blockchain.go:491]: I specifically kept it -1 to make it consistent with docs/comments.

**Comment by ceyonur** [comment]
closing in favor of #399
